runtime/src/bank.rs

```diff
@@ -3085,6 +3087,104 @@ impl Bank {
         }
     }
 
+    // TODO - solution 1, check vote redundancy by VoteState within bank,
+    //     this solution does not alight with bankless leader goal;
+    //     mainly using it to compare with solution #2 to confirm
+    //     behavior and performance.
+    fn is_redundant_vote_by_vote_state(&self, vote_account_pubkey: &Pubkey, vote: &Vote) -> bool {
+        // ignore vote without slot, or its last slot is '0'
+        let last_slot = match vote.slots.last() {
+            None => {
+                debug!(
+                    "Vote  has no slots, skip checking redundancy for vote {:?}",
+                    vote
+                );
+                return false;
+            }
+            Some(slot) => {
+                if 0 == *slot {
+                    debug!(
+                        "Vote on slot 0, skip checking redundancy for vote {:?}",
+                        vote
+                    );
+                    return false;
```

Comment:
Couldn't a later vote have already landed, why do we make an exception for slot 0?

Reply:
So when the validator starts up, I noticed the first votes are for slot 0, at that moment `last_voted_slot` is `0` too. Hence the special path for slot 0 to avoid it being dropped as redundant.